feat: Add built-in migration lock to prevent concurrent Prisma migrate deploy

minijeong-log opened the pull request:
Author: Mini Jeong <mini.jeong@navercorp.com>

## Title

<!-- e.g. "Implement user authentication feature" -->

Add built-in migration lock & startup gating to prevent multi-pod Prisma migrate deploy races

## Relevant issues

<!-- e.g. "Fixes #000" -->

Features #14400 

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem
<img width="2202" height="445" alt="screenshot 2025-09-11 5 50 55" src="https://github.com/user-attachments/assets/eccead33-5c89-4625-b545-d91a5669b222" />



## Type

<!-- Select the type of Pull Request -->
<!-- Keep only the necessary ones -->

🆕 New Feature

## Changes

### Summary

This PR adds an opt-in startup guardrail to serialize Prisma schema migrations across replicas. Exactly one pod acquires a database-scoped lock and runs prisma migrate deploy; other pods wait with backoff/timeout. This largely eliminates migration races, connection storms, and the downstream errors caused by serving against a partially-migrated schema.

### What’s in this PR

- Single-runner migration via DB advisory lock
   - On startup, a pod tries to acquire a DB-scoped advisory lock.
   - The lock holder runs prisma migrate deploy.
   - Non-holders wait until the lock is released or a failure/timeout occurs.
   - Clear logs for lock acquire/wait/release and migration result.

- Failure-handling & retry semantics
   - If the lock holder’s migration fails, it releases the lock and exits with a non-zero status.
   - Subsequent pods (or restarted pods) will contend for the lock; the next lock holder re-attempts prisma migrate deploy to completion (Prisma’s migrate deploy is designed to be safely re-runnable).
   - The advisory lock wraps every migration attempt—including failed ones—so even retries are serialized, preventing a thundering herd against the DB.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds a Redis-based migration lock to prevent concurrent Prisma migrations in multi-pod setups, plus wiring and tests.
> 
> - Introduce `MigrationLockManager` (Redis `SET NX` + TTL) with acquire/wait/release and context manager in `litellm_proxy_extras/utils.py`
> - Update `ProxyExtrasDBManager.setup_database(use_migrate, redis_cache)` to require `DATABASE_URL`, gate migrations with `MigrationLockManager`, and extract `_execute_migration`
> - `prisma_client.py` passes `redis_usage_cache` to `ProxyExtrasDBManager.setup_database` to enable distributed locking
> - Add comprehensive tests for lock behavior and migration flow in `tests/litellm-proxy-extras/test_litellm_proxy_extras_utils.py`
> - Minor typing cleanup in `vertex_llm_base.py` (use `cast` and unify `Literal` version ordering)
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit e94c9fe82648b982017c98263ae857f01d65e1c5. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

minijeong-log committed:
feat: prisma migrate deploy with lock
Author: Mini Jeong <mini.jeong@navercorp.com>

krrishdholakia commented on litellm/proxy/db/prisma_client.py:
```diff
@@ -154,7 +154,14 @@ def setup_database(use_migrate: bool = False) -> bool:
                     prisma_dir = PrismaManager._get_prisma_dir()
 
-                    return ProxyExtrasDBManager.setup_database(use_migrate=use_migrate)
+                    redis_cache = None
```
please type the variables correctly

krrishdholakia commented on litellm/proxy/db/prisma_client.py:
```diff
@@ -154,7 +154,14 @@ def setup_database(use_migrate: bool = False) -> bool:
+                    redis_cache = None
+                    try:
+                        from litellm.caching.redis_cache import RedisCache
+                        redis_cache = RedisCache()
```
the proxy already has an internal usage cache - see proxy_server.py for it. 

Use that instead.

minijeong-log replied:
I’ve updated the code to use the `redis_usage_cache` provided by the proxy server.

minijeong-log committed:
fix: use redis cache from proxy server
Author: Mini Jeong <mini.jeong@navercorp.com>

minijeong-log commented on litellm-proxy-extras/litellm_proxy_extras/utils.py:
```diff
@@ -17,6 +17,102 @@ def str_to_bool(value: Optional[str]) -> bool:
+    MIGRATION_LOCK_KEY = "migration_lock"
+    LOCK_TTL_SECONDS = 300  # 5 minutes TTL
+
+    def __init__(self, redis_cache=None):
```
I’ve updated the type to RedisCache.

minijeong-log committed:
Merge branch 'BerriAI:main' into feat/migration-lock

minijeong-log committed:
Merge branch 'BerriAI:main' into feat/migration-lock

minijeong-log committed:
fix: add type checks and fix unit tests for migration lock

- Add DATABASE_URL validation in _create_baseline_migration() and _resolve_all_migrations()
- Fix MyPy type errors by adding None checks before using database_url in subprocess calls
- Add _resolve_all_migrations mock to failing unit tests to prevent filesystem errors
- Apply Black formatting to modified files

Fixes:
- MyPy type errors: database_url could be None when passed to subprocess
- Unit test failures: _resolve_all_migrations tried to create directories in read-only /test path

minijeong-log committed:
fix: resolve MyPy type error in vertex_ai vertex_llm_base

Fix MyPy type checking error where vertex_api_version parameter type
was incompatible with function signature expectation.

nmgarza5 committed:
feat: add display_name, model_vendor, and model_version metadata

Eric84626 committed:
fix: Return 403 exception when calling GET responses api

Eric84626 committed:
Merge branch 'BerriAI:main' into main

Eric84626 committed:
fix: added new step into rotate master key function for processing credentials table

Sameerlite committed:
Add redisvl in requirements.txt

Eric84626 committed:
Merge branch 'BerriAI:main' into main

mangabits committed:
Use already configured opentelemetry providers

Users that instrument using opentelemetry-instrument can now setup exporters as per their environment.

Eric84626 committed:
fix: fixed the issue of handling root paths when processing Discovery protected resource metadata and authorization server metadata URLs.

Eric84626 committed:
fix: added additional grant type into oauth_authorization_server response for fixing mcp auth register bad request issue

Eric84626 committed:
Merge pull request #1 from Eric84626/feature/fix_metadata_url_auth_servers_url_with_rootpath

fix: fixed the issue of handling root paths when processing Discovery…

Eric84626 committed:
Merge pull request #2 from Eric84626/feature/add_grant_type_for_mcp_register_api

fix: added additional grant type into oauth_authorization_server resp…

Eric84626 committed:
fix: added RFC RECOMMENDED property(scopes_supported) to protected resource and authorization server metadata

Eric84626 committed:
Merge pull request #3 from Eric84626/feature/add_scopes_supported

fix: added RFC RECOMMENDED property(scopes_supported) to protected re…

Eric84626 committed:
fix: removed initialize the tool name to MCP server name mapping(oauth2) on startup for avoiding 401 error

Eric84626 committed:
Merge pull request #4 from Eric84626/feature/fix_mcp_401_error_startup_proxyserver

fix: removed initialize the tool name to MCP server name mapping(oaut…

Eric84626 committed:
fix: upgraded mcp sdk depency version for fixing ClosedResourceError

Eric84626 committed:
Merge pull request #5 from Eric84626/feature/fix_mcp_closed_resource_error

fix: upgraded mcp sdk depency version for fixing ClosedResourceError

LingXuanYin committed:
Align responses API streaming hooks with chat pipeline

mangabits committed:
Handle all protocols for all telemetry

mangabits committed:
Add more tests

LingXuanYin committed:
Clarify responses API streaming context

ktsaou committed:
feat(mcp): parallelize tool fetching from multiple MCP servers (#18627)

* feat(mcp): parallelize tool fetching from multiple MCP servers

Replace sequential tool fetching with asyncio.gather() to reduce
client timeouts when using multiple MCP servers.

Changes:
- mcp_server_manager.py: list_tools() now fetches tools in parallel
- server.py: _get_tools_from_mcp_servers() now fetches tools in parallel

Real-world impact (7 MCP servers example):
- Sequential: ~4.5+ seconds (exceeds typical 5-second client timeouts)
- Parallel: ~1.2 seconds (max of all servers)

Fixes #18626

* fix: copy oauth2_headers to avoid shared dict mutation in parallel tasks

shivamrawat1 committed:
added the option of adding langsmith tenant id in the env (#18623)

FlibbertyGibbitz committed:
fix(router): Validate routing_strategy at startup to fail fast with helpful error. (#18624)

Invalid routing_strategy values (e.g., "simple" instead of "simple-shuffle") previously failed silently, causing confusing "No deployments available" errors downstream. This change adds upfront validation in routing_strategy_init() to:

- Check if the provided strategy matches valid string values or RoutingStrategy enum
- Raise a clear ValueError listing valid options if invalid
- Fail fast at startup instead of at request time

Fixes behavior reported in #11330 where users had to debug cryptic errors.

Valid strategies: simple-shuffle, least-busy, usage-based-routing, latency-based-routing, cost-based-routing, usage-based-routing-v2

Co-authored-by: Flibbert E. Gibbitz <flibbertygibbitz@runelabs.ai>

Chesars committed:
Add libsndfile to database Docker image for audio processing (#18612)

The litellm-database Docker image was missing the libsndfile system
library, which is required by the soundfile Python package for audio
file processing. This caused failures when using audio transcription
endpoints that attempt to calculate audio duration.

This adds libsndfile to the runtime dependencies in Dockerfile.database,
consistent with Dockerfile.alpine which already includes this library.

Chesars committed:
Fix: Map Gemini cached_tokens to Langfuse cache_read_input_tokens (#18614)

* Fix: Map Gemini cached_tokens to Langfuse cache_read_input_tokens

Fixes #18520

## Problem
Langfuse integration was not capturing cached tokens from Gemini models.
Gemini returns cached tokens in `usage.prompt_tokens_details.cached_tokens`,
but Langfuse only read from top-level `usage.cache_read_input_tokens`
(which only Anthropic populates).

## Solution
Updated langfuse.py to check both locations:
1. First check top-level cache_read_input_tokens (for Anthropic)
2. Then check prompt_tokens_details.cached_tokens (for Gemini, OpenAI, others)

This ensures all providers' cached tokens are properly reported to Langfuse.

## Changes
- Modified litellm/integrations/langfuse/langfuse.py (lines 742-761)
- Added 3 unit tests in tests/test_litellm/integrations/langfuse/test_gemini_cached_tokens.py
- All existing Langfuse tests still pass (11/11)

## Testing
- test_cached_tokens_extraction: Verifies Gemini cached_tokens extraction
- test_cached_tokens_not_present: Backward compatibility (no cached_tokens)
- test_cached_tokens_is_zero: Edge case when cached_tokens = 0

* Refactor: Extract cache token logic into helper function

Address review feedback from @officer47p

- Created _extract_cache_read_input_tokens() helper function
- Reduces code bloat in _log_langfuse_v2 method
- Improves testability and reusability
- All tests still passing (11/11)

Eric84626 committed:
Merge branch 'main' into main

yuneng-jiang committed:
Adding Role Mappings

yuneng-jiang committed:
Merge pull request #18677 from BerriAI/litellm_ui_sso_settings_role

[Feature] UI - SSO Settings Page Add Role Mappings

yuneng-jiang committed:
Fixing Edit SSO Settings Modal

uc4w6c committed:
feat: add user_mcp_management_mode for view_all visibility

yuneng-jiang committed:
Fixing tests

uc4w6c committed:
fix: missing mcp_allow_all_ui.png

uc4w6c committed:
docs: add user_mcp_management_mode

mangabits committed:
Address review comments

uc4w6c committed:
Merge pull request #18279 from mangabits/fix-otel-provider

Use already configured opentelemetry providers

uc4w6c committed:
Merge pull request #18681 from BerriAI/litellm_feat_mcp-server-visibility-configurable

[feat] mcp server visibility configurable

yuneng-jiang committed:
Merge pull request #18680 from BerriAI/litellm_sso_modal_fix

[Fix] UI - SSO Edit Modal Clear Role Mapping Values on Provider Change

uc4w6c committed:
Merge pull request #18281 from Eric84626/main

Fix: fixed some MCP gateway oauth2 auth issues and ClosedResourceError

0717376 committed:
feat: Add GigaChat provider support (#18564)

* feat: Add GigaChat provider support

Add native support for GigaChat API (Sber AI, Russia's leading LLM).

Supported features:
- Chat completions (sync/async)
- Streaming (sync/async)
- Function calling / Tools
- Structured output via JSON schema (emulated through function calls)
- Image input (base64 and URL)
- Embeddings

Closes #18515

* fix: resolve mypy type errors in GigaChat handler

- Fix _prepare_file_data return type (use 3-tuple for cleaner type flow)
- Add type annotations for lists in _process_content_parts methods
- Add type annotations in _collapse_user_messages
- Use ChatCompletionToolCallChunk for proper tool_use typing
- Add type: ignore[override] for astreaming async generator

* refactor(gigachat): migrate to BaseConfig pattern

* fix: remove unused imports

* fix: resolve mypy type errors

* fix: mypy type errors

* refactor: address review feedback for GigaChat provider

- Remove singleton pattern, reuse litellm HTTPHandler
- Move constants/errors to transformation files, delete common_utils.py
- Add models to model_prices_and_context_window.json
- Fix ssl_verify not passed to HTTP client for embeddings

* docs: update GigaChat documentation with ssl_verify requirement

Sameerlite committed:
Merge pull request #18177 from BerriAI/litellm_fix_redisvl_dependency

Add redisvl in requirements.txt

Sameerlite committed:
Merge pull request #18663 from BerriAI/litellm_staging_01_05_2026

Staging 01/05/2026

Sameerlite committed:
Merge pull request #18513 from talesofai/feat/responses-api-hooks-parity

feat (proxy) : Responses API Support More Streaming Callback Hooks

Sameerlite committed:
Merge pull request #17330 from nmgarza5/add-model-metadata

feat(model_cost): add display_name, model_vendor, and model_version metadata to model entries

uc4w6c committed:
Revert "Add redisvl in requirements.txt"

Sameerlite committed:
Put reasoning summary behind feat flag

uc4w6c committed:
Merge pull request #18687 from BerriAI/revert-18177-litellm_fix_redisvl_dependency

Revert "Add redisvl in requirements.txt"

uc4w6c committed:
fix: model eol

uc4w6c committed:
fix: anthropic claude-3-opus-20240229 EOL

uc4w6c committed:
Revert "fix: model eol"

This reverts commit 5aa1665d79d75e0842ec44a8dc23d2e15fdfadd8.

uc4w6c committed:
Merge pull request #18693 from BerriAI/litellm_fix_main_test

fix main test

Sameerlite committed:
Fix: ImportError: qualifire package is required for QualifireGuardrail. Install it with: pip install qualifire

Sameerlite committed:
fix: test_secret_manager_failure_does_not_block_email

Sameerlite committed:
fix: test_update_ui_settings_allowlisted_value

Sameerlite committed:
fix: test_aaamodel_prices_and_context_window_json_is_valid

Sameerlite committed:
fix: test_all_models_have_display_name

Sameerlite committed:
fix: async def test_bedrock_apply_guardrail_blocked()

Sameerlite committed:
fix: test_databricks_embeddings[True]

Sameerlite committed:
fix:test_anthropic_beta_header

Sameerlite committed:
fix:test_api_error_handling

Sameerlite committed:
fix:mypy mcp management

Sameerlite committed:
Merge pull request #18695 from BerriAI/litellm_fix_mapped_test_060126

Fix mapped test 060126

Sameerlite committed:
Revert "feat(model_cost): add display_name, model_vendor, and model_version metadata to model entries"

Sameerlite committed:
Merge pull request #18701 from BerriAI/revert-17330-add-model-metadata

Revert "feat(model_cost): add display_name, model_vendor, and model_version metadata to model entries"

ishaan-jaff committed:
[Feat] New API Endpoint - Responses API (v1/responses/compact) (#18697)

* init transform_compact_response_api_request

* init acompact_responses

* init async_compact_response_api_handler in llm http handler

* init transform_compact_response_api_request for openai

* init acompact_responses

* fix acompact_responses

* add OAI Compact API

* docs responses API Compact

* code qa checks

* test_openai_compact_responses_api

* fix mypy linting

Sameerlite committed:
fix: remove display name

Sameerlite committed:
Merge pull request #18702 from BerriAI/litellm_fix_mapped_test_3

fix: remove display name

Sameerlite committed:
Merge pull request #18688 from BerriAI/litellm_fix_reasoning_summary_default

[Fix] Put automatic reasoning summary inclusion behind feat flag

Sameerlite committed:
Add the LITELLM_REASONING_AUTO_SUMMARY in doc

Sameerlite committed:
Merge pull request #18705 from BerriAI/litellm_fix_reasoning_summary_default

Add the LITELLM_REASONING_AUTO_SUMMARY in doc

Sameerlite committed:
fix model map

Sameerlite committed:
Merge pull request #18706 from BerriAI/litellm_fix_reasoning_summary_default

[Bug] fix model map

ishaan-jaff committed:
[UI] - Feat add request provider form on UI (#18704)

* add request provider form

* fix link to github

* add button

* fix link

krisxia0506 committed:
fix(streaming): normalize status code extraction to prevent 4xx errors from triggering mid-stream fallback (#18698)

在流式处理错误时，添加状态码标准化逻辑，确保 4xx 客户端错误直接抛出而不是被包装成 MidStreamFallbackError。

- 新增 _normalize_status_code 函数用于从异常对象提取状态码
- 优先从异常的 status_code 属性获取，其次从 response.status_code 获取
- 当映射异常或原始异常的状态码在 400-499 范围内时，直接抛出映射异常
- 添加单元测试验证 Vertex AI 400 错误正确抛出为 BadRequestError
- 确保流式处理中的客户端错误能够正确传播，而不会触发回退机制

Commit:
Merge branch 'main' into feat/migration-lock

krrishdholakia commented:
Had to revert the PR as it led to 87 file changes on staging. 

@minijeong-log can you resubmit with the updated codebase?

krrishdholakia commented:
@Harshit28j can you please take over this pr?
